scripts/sap_automation_qa.sh

```diff
@@ -100,24 +147,155 @@ get_playbook_name() {
     esac
 }
 
-# Function to run the ansible playbook
+"""
+Retrieve a secret from Azure Key Vault.
+
+:param key_vault_id: The ID of the Key Vault.
+:return: None. Exits with a non-zero status if retrieval fails.
+"""
+retrieve_secret_from_key_vault() {
+    local key_vault_id=$1
+    local required_permission="Get"
+
+    # Extract resource group name and key vault name from key_vault_id
+    resource_group_name=$(echo "$key_vault_id" | awk -F'/' '{for(i=1;i<=NF;i++){if($i=="resourceGroups"){print $(i+1)}}}')
+    key_vault_name=$(echo "$key_vault_id" | awk -F'/' '{for(i=1;i<=NF;i++){if($i=="vaults"){print $(i+1)}}}')
+    subscription_id=$(echo "$key_vault_id" | awk -F'/' '{for(i=1;i<=NF;i++){if($i=="subscriptions"){print $(i+1)}}}')
+
+    if [[ -z "$resource_group_name" || -z "$key_vault_name" ]]; then
+        log "ERROR" "Failed to extract resource group name or key vault name from key_vault_id: $key_vault_id"
+        exit 1
+    fi
+
+    log "INFO" "Extracted resource group name: $resource_group_name"
+    log "INFO" "Extracted key vault name: $key_vault_name"
+
+    # Authenticate using MSI
+    log "INFO" "Authenticating using MSI..."
+    az login --identity
+    az account set --subscription "$subscription_id"
+    if [[ $? -ne 0 ]]; then
+        log "ERROR" "Failed to authenticate using MSI."
+        exit 1
+    fi
+
+    # Attempt to access Key Vault to verify permissions
+    log "INFO" "Verifying permissions on Key Vault: $key_vault_name..."
+    set +e  # Temporarily disable exit on error
+    error_message=$(az keyvault secret list --vault-name "$key_vault_name" 2>&1)
+    az_exit_code=$?  # Capture the exit code of the az command
+    set -e  # Re-enable exit on error
+
+    if [[ $az_exit_code -ne 0 ]]; then
+        extracted_message=$(extract_error_message "$error_message")
+        log "ERROR" "Azure CLI error: $extracted_message"
+        exit 1
+    fi
+    
+    # Attempt to retrieve the secret value and handle errors
+    log "INFO" "Retrieving secret '$secret_name' from Key Vault '$key_vault_name'..."
+    set +e  # Temporarily disable exit on error
+    secret_value=$(az keyvault secret show --vault-name "$key_vault_name" --name "$secret_name" --query "value" -o tsv 2>&1)
```

Comment:
this will only ensure that the std error is redirected to std output
please add > /dev/null to discard std output